CI: Add rootfs definitions for `buildkite-agent`, `package_linux64`, and `package_musl64`

staticfloat opened the pull request:
This adds an Alpine-based rootfs image for packaging jobs on `musl64`.
It purposefully installs GCC 9, expecting that we will standardize on GCC 9 as our buildbot GCC version for as many platforms as possible.

staticfloat committed:
Add definition for `package_musl64` rootfs image

This adds an Alpine-based rootfs image for packaging jobs on `musl64`.
It purposefully installs GCC 9, expecting that we will standardize on
GCC 9 as our buildbot GCC version for as many platforms as possible.

staticfloat commented:
I'm marking this as a draft until we're satisfied with its setup.

ararslan commented on .buildkite/rootfs_images/rootfs_utils.jl:
```diff
@@ -74,12 +88,59 @@ function debootstrap(name::String; release::String="buster", variant::String="mi
+                apk_cmd = `$(apk_cmd) $(package.name)`
+            end
+            run(`sudo chroot --userspec=$(getuid()):$(getgid()) $(rootfs) $(apk_cmd)`)
+        end        
```
```suggestion
        end
```
This line is what's making the whitespace check sad

ararslan commented on .buildkite/rootfs_images/rootfs_utils.jl:
```diff
@@ -74,12 +88,59 @@ function debootstrap(name::String; release::String="buster", variant::String="mi
-    return tarball_path
+    AlpinePackage(name, repo=nothing) = new(name, repo)
+end
+function repository_arg(repo)
+    if repo === nothing
+        return String[]
+    end
+    if startswith(repo, "https://")
+        return ["--repository=$(repo)"]
+    end
+    return ["--repository=http://dl-cdn.alpinelinux.org/alpine/$(repo)/main"]
+end
```
```suggestion
function repository_arg(repo)
    if startswith(repo, "https://")
        return ["--repository=$(repo)"]
    end
    return ["--repository=http://dl-cdn.alpinelinux.org/alpine/$(repo)/main"]
end
repository_arg(::Nothing) = String[]
```

KristofferC replied:
I think nothing checks like this tend to be inline in the main function.

staticfloat committed:
Add `buildkite-agent` rootfs image

This contains the rootfs that we use to run the actual `buildkite-agent`
process within.  It's much more minimal than most of these rootfs
images, and is purposefully so, to encourage CI passes to be run within
a pass-specific rootfs image.

This also beefs up the rootfs construction utilities somewhat,
incorporating some of the best practices scattered around various
similar efforts.

staticfloat committed:
Add `test_rootfs.jl`

This is pretty useful for quickly testing out whether a `rootfs` works
or not.  Use via `julia test_rootfs.jl <url> <hash>`.

staticfloat committed:
Add `package_linux64` rootfs image

This rootfs image contains the necessary tools to build Julia for glibc
x86_64 linux, using GCC 9.

staticfloat committed:
Always initialize all CI images with a `juliaci` user and home directory

staticfloat committed:
Integrate changes from upstream `master`

DilumAluthge commented:
For building 32-bit Julia, i.e. `package_linux32`, will we need a separate `package_linux32` rootfs image?

DilumAluthge committed:
Fix whitespace

DilumAluthge commented:
It would be good (for reproducibility purposes) to also add the rootfs definition for `pkgserver-logsync.tar.gz`. It doesn't need to be done in this PR.

DilumAluthge committed:
Add instructions for building the rootfs images with Docker

DilumAluthge committed:
Delete the Docker instructions - they may lead to broken rootfs images

DilumAluthge commented:
I'm closing this PR in favor of moving to to the [JuliaCI/rootfs-images](repository).

I've broken it up into several smaller PRs so that we can merge the different images independently of each other. Track the progress here: https://github.com/JuliaCI/rootfs-images/issues/6

DilumAluthge commented:
> It purposefully installs GCC 9, expecting that we will standardize on GCC 9 as our buildbot GCC version for as many platforms as possible.

@staticfloat Out of curiosity, what are the reasons for selecting GCC 9 as our standard GCC version?
